src/celeritas/optical/surface/VolumeSurfaceSelector.hh

```diff
@@ -71,69 +75,39 @@ class VolumeSurfaceSelector
 /*!
  * Construct with pre-volume IDs.
  */
-CELER_FUNCTION VolumeSurfaceSelector::VolumeSurfaceSelector(
-    NativeCRef<SurfaceParamsData> const& params,
-    VolumeId pre_volume,
-    VolumeInstanceId pre_volume_inst)
-    : params_(params)
-    , pre_volume_(pre_volume)
-    , pre_volume_inst_(pre_volume_inst)
+CELER_FUNCTION
+VolumeSurfaceSelector::VolumeSurfaceSelector(VolumeSurfaceView pre_surface,
+                                             VolumeInstanceId pre_volume_inst)
+    : pre_surface_(std::move(pre_surface)), pre_volume_inst_(pre_volume_inst)
 {
-    CELER_EXPECT(pre_volume_ < params_.volume_surfaces.size());
     CELER_EXPECT(pre_volume_inst_);
 }
 
-//---------------------------------------------------------------------------//
-/*!
- * Convenience constructor to use IDs from geometry.
- *
- * FIXME: use canonical volume + instance
- */
-CELER_FUNCTION VolumeSurfaceSelector::VolumeSurfaceSelector(
-    NativeCRef<SurfaceParamsData> const& params, GeoTrackView const& geo)
-    : VolumeSurfaceSelector(
-          params, geo.impl_volume_id(), geo.volume_instance_id())
-{
-    CELER_EXPECT(!geo.is_outside());
-}
-
 //---------------------------------------------------------------------------//
 /*!
  * Select surface based on post-volume IDs.
  *
  * Returns an invalid \c SurfaceId if no surface data exists for the volumes.
  */
-CELER_FUNCTION SurfaceId VolumeSurfaceSelector::operator()(
-    VolumeId post_volume, VolumeInstanceId post_volume_inst) const
+CELER_FUNCTION auto
+VolumeSurfaceSelector::operator()(VolumeSurfaceView const& post_surface,
+                                  VolumeInstanceId post_volume_inst) const
+    -> OrientedSurface
 {
-    VolumeSurfaceView pre_surface{params_, pre_volume_};
-
     if (auto surface_id
-        = pre_surface.find_interface(pre_volume_inst_, post_volume_inst))
+        = pre_surface_.find_interface(pre_volume_inst_, post_volume_inst))
     {
-        return surface_id;
+        return {surface_id, SubsurfaceDirection::forward};
     }
-
-    if (auto surface_id = pre_surface.boundary_id())
+    else if (auto surface_id = pre_surface_.boundary_id())
     {
-        return surface_id;
+        return {surface_id, SubsurfaceDirection::forward};
+    }
+    else
+    {
+        // If no boundary in post-volume then return invalid SurfaceId
+        return {post_surface.boundary_id(), SubsurfaceDirection::reverse};
```

Comment:
At first this comment sounds like it's inconsistent with the code but I get what you mean. Maybe
```suggestion
        // Return volume boundary from the opposite direction, which is just a "false"
        // OrientedSurface if there is no post-volume boundary.
        return {post_surface.boundary_id(), SubsurfaceDirection::reverse};
```